statuspage/index.js

```diff
@@ -0,0 +1,135 @@
+import express from "express";
+import { readFile } from "fs/promises";
+import http from "http";
+import https from "https";
+import interval from "interval-promise";
+import moment from "moment-timezone";
+
+/**
+ * this service lives outside of Kubernetes
+ *
+ * because this exists independent of Kubernetes, the cert loading process is
+ * manual
+ *
+ * to manually pull certificates from k8s:
+ * kubectl get secrets/bt-tls --template='{{index .data "tls.crt"}}' | base64 --decode > bt-tls.cert
+ * kubectl get secrets/bt-tls --template='{{index.data "tls.key"}}' | base64 --decode > bt-tls.key
+ *
+ * to create a DNS A record for status.berkeleytime.com
+ * export ADDRESS=34.82.91.206
+ * gcloud dns record-sets transaction start --project berkeleytime-218606 --zone berkeleytime
+ * gcloud dns record-sets transaction add --project berkeleytime-218606 --zone berkeleytime --name berkeleytime.com --ttl 300 --type A $ADDRESS
+ * gcloud dns record-sets transaction execute --project berkeleytime-218606 --zone berkeleytime
+ *
+ * TODO:
+ * pull certificate + key every x amount of days and then restart server
+ */
+const cert = await readFile("bt-tls.cert");
+const key = await readFile("bt-tls.key");
+
+import {
+  PORT_EXPRESS_HTTP,
+  PORT_EXPRESS_HTTPS,
+  STATUS_DEGRADED_PERFORMANCE,
+  STATUS_OPERATIONAL,
+  STATUS_MAJOR_OUTAGE,
+} from "./config.js";
+import spConfig from "./config.statuspage.json";
+
+axios.defaults.headers["Authorization"] = `OAuth ${spConfig.apiKey}`;
+
+/* update berkeleytime.statuspage.io **************************************** */
+async function updateStatuspage(ping, auth) {
+  const { pageId, metridId } = spConfig;
+  const { componentId, metricId } = auth;
+  const statusUrl = `https://api.statuspage.io/v1/pages/${pageId}`;
+
+  try {
+    let start = Date.now();
+    await ping();
+    let responseTime = Date.now() - start;
+
+    await axios.post(`${statusUrl}/metrics/${metricId}/data.json`, {
+      data: {
+        timestamp: Math.floor(start / 1000),
+        value: responseTime,
+      },
+    });
+
+    let status;
+    if (responseTime > 2000) {
+      status = STATUS_DEGRADED_PERFORMANCE;
+    } else {
+      status = STATUS_OPERATIONAL;
+    }
+
+    await axios.patch(`${statusUrl}/components/${componentId}`, {
+      component: {
+        status,
+      },
+    });
+  } catch (err) {
+    console.error(err);
+    await axios.patch(`${statusUrl}/components/${componentId}`, {
+      component: {
+        status: STATUS_MAJOR_OUTAGE,
+      },
+    });
+  }
+}
+
+// REST API check
+// await updateStatuspage(async () => {
+//   await axios.get(
+//     `berkeleytime.com/api/catalog/catalog_json/course_box/?course_id=2321`
+//   );
+// }, spConfig.restStatus);
+
+interval(async () => {
+  const { apiStatus, statusInterval, frontendStatus, url } = spConfig;
+  await updateStatuspage(async () => {
+    await axios.post(`${url}/api/graphql`, {
+      query: "query PingQuery{ ping }",
+    });
```

Comment:
I think I had some checks for the response status code as well as if the response data was valid JSON, cuz sometimes (at least the REST API) I think it defaults to sending some sort of HTML/XML when the backend is not working. Similar thing for the frontend.

Reply:
The reason I wanted to remove those is because axios.post and axios.get are going to throw errors if the status isn't 200, so I believe those checks wouldn't be necessary, since you had already implemented catch logic on line 72-80

```
await axios.post
if (statusCode = 200 && response.data == "blah") {
... this block will never hit if statusCode is not 200 ...
... therefore only necessary if it's possible for status 200 to come back but still be failure case ...
}
```

Reply:
I see. sounds good to me then

Reply:
![image](https://user-images.githubusercontent.com/22272118/114600295-0677eb00-9c49-11eb-8ff3-e58472f9abf1.png)

the above test was for when I deleted the API key, but actually it's worth testing out a real backend failure then

Reply:
@kevinzwang confirmed
![image](https://user-images.githubusercontent.com/22272118/114601180-104e1e00-9c4a-11eb-9f82-a90510d573ae.png)
![image](https://user-images.githubusercontent.com/22272118/114601107-f90f3080-9c49-11eb-8e82-bc36db403326.png)
![image](https://user-images.githubusercontent.com/22272118/114601110-fa405d80-9c49-11eb-8620-93ce04e5d570.png)
![image](https://user-images.githubusercontent.com/22272118/114601111-fad8f400-9c49-11eb-8864-a1fec6834ca2.png)
![image](https://user-images.githubusercontent.com/22272118/114601288-3247a080-9c4a-11eb-8b81-8c26020488fb.png)